lua/lspconfig/ui/lspinfo.lua

```diff
@@ -214,24 +260,53 @@ return function()
     end
   end
 
-  local matching_config_header = {
-    '',
-    'Configured servers list: ' .. table.concat(vim.tbl_keys(configs), ', '),
+  local configured_head = 'Configured servers list: '
+  local start_row = #buf_lines
+  local servers_ctx = {
+    servers = util.available_servers(),
+    width = floating_width,
+    header = configured_head,
   }
-  vim.list_extend(buf_lines, matching_config_header)
+  local servers_list, hi_scope = generate_servers_list(servers_ctx)
+
+  vim.list_extend(buf_lines, servers_list)
 
   local fmt_buf_lines = indent_lines(buf_lines, ' ')
 
   fmt_buf_lines = vim.lsp.util._trim(fmt_buf_lines, {})
 
-  vim.api.nvim_buf_set_lines(bufnr, 0, -1, true, fmt_buf_lines)
-  vim.api.nvim_buf_set_option(bufnr, 'modifiable', false)
-  vim.api.nvim_buf_set_option(bufnr, 'filetype', 'lspinfo')
+  api.nvim_buf_set_lines(bufnr, 0, -1, true, fmt_buf_lines)
+  api.nvim_buf_set_option(bufnr, 'modifiable', false)
+  api.nvim_buf_set_option(bufnr, 'filetype', 'lspinfo')
 
-  vim.api.nvim_buf_set_keymap(bufnr, 'n', '<esc>', '<cmd>bd<CR>', { noremap = true })
-  vim.api.nvim_command(
-    string.format('autocmd BufHidden,BufLeave <buffer> ++once lua pcall(vim.api.nvim_win_close, %d, true)', win_id)
-  )
+  api.nvim_buf_attach(bufnr, false, {
+    on_detach = function()
+      if api.nvim_win_is_valid(win_id) then
+        api.nvim_win_close(win_id, true)
+      end
+    end,
+  })
```

Comment:
Why not `BufDelete` event instead?

Reply:
this not other side effect no need to remove the autocmd id I think

Reply:
Buffer autocommands should be automatically deleted when a buffer is deleted, but you can also add `once = true` to make sure.

Reply:
Nope, sorry, only when buffer is wiped out, not deleted. But still, `once = true`.

Alternatively you could use an augroup, and `nvim_clear_autocmds({ buffer = ..., group = ... })`

Reply:
yup . I'm curious about the difference between the BufDelete  nvim_buf_attatch?

Reply:
looks like use `nvim_buf_attach` most easy .can save some code.

Reply:
`on_detach` triggers mainly on :bunload, so this would be more equivalent to `BufUnload`.

Reply:
hmmm use `BufDelete` more save code .just add it into that events...

Reply:
on_detach also triggerd by `bwipeout`  :)

Reply:
Prefer autocommands, nvim_buf_attach is for listening for buffer updates. And `BufUnload` is a too high threshold, it should be `BufDelete`.
Actually, I think this can just added to that `BufLeave,BufHidden` autocommand.

Reply:
yup had finished. Okay I learned more about `nvim_buf_attach` thanks